### ISSUE TITLE
Promise & async/await support

### ISSUE THREAD

**Issue by lolleko** [TypeScriptToLua/TypeScriptToLua#147]
Add support for es6 promises. Probably via custom lib (in typescript_lualib.lua)

**Comment by tjahn** [comment]
I saw this request and just want to refer my (closed) PR #380.

The changes to translate the javascript keywords async and await to function calls (__TS__Async and __TS__Await) as well as the lua-keyword then "xyz.then" to "xyz.__TS__then" will not work anymore and will have to be rewritten.

However, the PR adds these three lualib files that include ES6 promises and two functions, that exactly mimc javascripts async and await keywords.
* Promise.ts
* Async.ts
* Await.ts

In addition some minor agreements were missing (see #380) as well as some tests

**Comment by weinixuehao** [comment]
```
src/lualib/Async.ts:16:33 - error TSTL: Unable to convert function with a 'this' parameter to function 'f' with no 'this'. To fix, wrap in an arrow function, or declare with 'thi.

16     const co = coroutine.create(f);
                                   ~~~~~~~~~~~~

src/lualib/Async.ts:22:24 - error TS6133: 'T' is declared but its value is never read.

22     function handleNext<T>(...args): any {
                          ~~~

src/lualib/Await.ts:23:5 - error TS2322: Type 'any[]' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to 'any[]'.

23     return result;
       ~~~~~~~~~~~~~~

src/lualib/Promise.ts:6:4 - error TS2304: Cannot find name 'print'.

6   (print as any)( "Unhandled rejected promise ", reason );
```
